addons/mail/controllers/discuss.py

```diff
@@ -1,18 +1,221 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-import base64
-import werkzeug.utils
-import werkzeug.wrappers
+import json
 
 from odoo import http
-from odoo.exceptions import AccessError
 from odoo.http import request
+from odoo.tools import consteq
+from odoo.tools.misc import get_lang
+from odoo.tools.translate import _
+from werkzeug.exceptions import NotFound
 
 
 class DiscussController(http.Controller):
 
-    @http.route('/mail/read_followers', type='json', auth='user')
+    # --------------------------------------------------------------------------
+    # Public Pages
+    # --------------------------------------------------------------------------
+
+    @http.route('/chat/<int:channel_id>/<string:invitation_token>', methods=['GET'], type='http', auth='public')
+    def discuss_channel_invitation(self, channel_id, invitation_token, **kwargs):
+        channel_sudo = request.env['mail.channel'].browse(int(channel_id)).sudo().exists()
+        if not channel_sudo or not channel_sudo.uuid or not consteq(channel_sudo.uuid, invitation_token):
+            raise NotFound()
+        response = request.redirect(f'/discuss/channel/{channel_sudo.id}')
+        if request.env['mail.channel.partner']._get_as_sudo_from_request(request=request, channel_id=int(channel_id)):
+            return response
+        if request.session.uid:
+            channel_sudo.add_members([request.env.user.partner_id.id])
+            return response
+        guest = request.env['mail.guest']._get_guest_from_request(request)
+        if not guest:
+            guest = request.env['mail.guest'].sudo().create({
+                'country_id': request.env['res.country'].sudo().search([('code', '=', request.session.get('geoip', {}).get('country_code'))], limit=1).id,
+                'lang': get_lang(request.env).code,
+                'name': _("Guest"),
+                'timezone': request.env['mail.guest']._get_timezone_from_request(request),
+            })
+            response.set_cookie('mail.guest_id', str(guest.id))
+            response.set_cookie('mail.guest_access_token', guest.access_token)
```

Comment:
those are meant to be session cookies? can this be explicitly documented?
also, could we make them `httponly=True` since they never need to be read from the JS side

ideally they should be `secure=True` too, but we cannot 100% force HTTPS-only deployments, I think

Reply:
Indeed, those can be considered like session cookie for guest. Maybe some day we can merge it with the real session and maybe also with website visitor? But it's out of scope for now.

Good idea to add `httponly=True`, we don't need those cookies in JS.

Reply:
> Indeed, those can be considered like session cookie for guest. Maybe some day we can merge it with the real session and maybe also with website visitor? But it's out of scope for now.

Wait, I meant "session cookie" as in "non-permanent cookies that will be discarded if you close your browser". Because they have no expiration date. This is different from typical expiration for our "session_id" cookie ;-)

Reply:
Then I guess we don't want them to be "session", if the guest is closing his browser it would be convenient to restore his identity instead of creating a new one. So we would want those cookie to live as long as possible.

Is there anything to change here then? In code or in the comment I added?

Reply:
> Is there anything to change here then? In code or in the comment I added?

Well yes, you have to set an expiration date or max-age on the cookie, that's how you make it permanent on the browser side. For `session_id` we put the validity [at 3 months](https://github.com/odoo/odoo/blob/master/odoo/http.py#L1429), but we auto-renew the cookie frequently.
For `visitor_uuid` we use a fixed [1 year validity](https://github.com/odoo/odoo/blob/master/addons/website/models/website_visitor.py#L205-L206).

So it depends: how long should we keep the same "guest"? 3 months or 1 year seems plausible values.

Then just pass an appropriate max_age value to set_cookie, as in the examples.